hcn/hcnsupport.go

```
@@ -45,54 +50,51 @@ type ApiSupport struct {
-	features.AclSupportForProtocol252 = isFeatureSupported(globals.Version, AclSupportForProtocol252Version)
-	features.SessionAffinity = isFeatureSupported(globals.Version, SessionAffinityVersion)
-	features.IPv6DualStack = isFeatureSupported(globals.Version, IPv6DualStackVersion)
-	features.SetPolicy = isFeatureSupported(globals.Version, SetPolicyVersion)
-	features.VxlanPort = isFeatureSupported(globals.Version, VxlanPortVersion)
-	features.L4Proxy = isFeatureSupported(globals.Version, L4ProxyPolicyVersion)
-	features.L4WfpProxy = isFeatureSupported(globals.Version, L4WfpProxyPolicyVersion)
-	features.TierAcl = isFeatureSupported(globals.Version, TierAclPolicyVersion)
-	features.NetworkACL = isFeatureSupported(globals.Version, NetworkACLPolicyVersion)
-	features.NestedIpSet = isFeatureSupported(globals.Version, NestedIpSetVersion)
-
-	// Only print the HCN version and features supported once, instead of everytime this is invoked. These logs are useful to
-	// debug incidents where there's confusion on if a feature is supported on the host machine. The sync.Once helps to avoid redundant
-	// spam of these anytime a check needs to be made for if an HCN feature is supported. This is a common occurrence in kubeproxy
-	// for example.
-	versionOnce.Do(func() {
-		logrus.WithFields(logrus.Fields{
-			"version":           fmt.Sprintf("%+v", globals.Version),
-			"supportedFeatures": fmt.Sprintf("%+v", features),
-		}).Info("HCN feature check")
+	// Only fetch the supported features and print the HCN version and features once, instead of everytime this is invoked.
+	// The logs are useful to debug incidents where there's confusion on if a feature is supported on the host machine. The sync.Once
+	// helps to avoid redundant spam of these anytime a check needs to be made for if an HCN feature is supported. This is a common
+	// occurrence in kubeproxy for example.
+	featuresOnce.Do(func() {
```

Consider a future failure mode introduced into this function. It might leave the global supportedFeatures in a partial state. Would be safter to operate on a local and move to the global at the end.

---

We'd talked offline but I think it would be better to have a version of this function that returns the error from `GetGlobals` if the query fails, instead of logging it. I don't feel like breaking existing callers though, so possibly a new function `SupportedFeatures()` that has the same logic but just returns the error (if there was one)